# inDAGO
![favicon-96x96](https://github.com/user-attachments/assets/4d325017-c452-4e8d-95a7-12f1f7ccd6f7)

A Shiny app for dual and bulk RNA‚Äësequencing analysis

## üëÄ Overview

**inDAGO** supports both dual and bulk RNA‚Äëseq workflows in a single, user-friendly Shiny interface. For dual RNA‚Äëseq, you can choose between:

- **Sequential mapping**: map reads separately to each reference genome  
- **Combined mapping**: align reads to a single, merged reference genome

The interface walks you step‚Äëby‚Äëstep through the entire analysis, from raw reads to publication‚Äëready plots, and lets you:

- Download intermediate results at each step  
- Export high‚Äëquality figures directly for your manuscript  

Thanks to optimized, parallelized code, inDAGO runs efficiently on a standard laptop (16‚ÄØGB RAM), so you don‚Äôt need access to a high‚Äëperformance cluster.

## üîß Key Modules

1. **Quality Control**  
   Generates quality control metrics and graphical plots.

   <details>
     <summary><strong>üìä See plots</strong> ‚ñ∏</summary>

     <br>

     ![QC](https://github.com/user-attachments/assets/4d55ecb6-6277-4971-bd4e-e25fe0e0eb29)

     <p align="justify">
     <em><strong>Figure:</strong> Quality Control Module Outputs.</em>  
     This figure presents key quality control plots generated by <strong>inDAGO</strong>:  
     (A) average base quality line plot;  
     (B) sequence length distribution;  
     (C) GC content distribution across reads;  
     (D) base quality boxplot showing average and variation per base position;  
     (E) base composition line plot; and  
     (F) base composition area chart across the dataset.  
     Together, these visualizations offer a comprehensive assessment of the quality and characteristics of the raw sequencing data.
     </p>

   </details>

  
3. **Sequence Pre‚Äëprocessing**  
   Read trimming, low‚Äëquality filtering, and adapter removal
4. **Genome indexing**  
   Index genome or genomes according to the selected approach (bulk or dual RNA‚Äëseq)
5. **Reference‚Äëbased Alignment**  
   Align reads according to the selected approach (bulk or dual RNA‚Äëseq) 
6. **Read Count Summarization**  
   Generate gene or transcript level count matrices  
7. **Exploratory Data Analysis**  
   PCA, MDS, heatmaps and more  
8. **Differential Expression Gene (DEG) analysis**  
   Identify differentially expressed genes/transcripts across comparisons

<details>
  <summary><strong>üíª INSTALLATION GUIDE: R AND RSTUDIO</strong> ‚ñ∏</summary>

###  1. Install R

**Official site:** [CRAN R Project](https://cran.r-project.org/)

| OS | Command or Link |
|----|-----------------|
| **Windows** | [Download R for Windows](https://cran.r-project.org/bin/windows/base/) and run the `.exe` installer.
| **macOS** | [Download R for macOS](https://cran.r-project.org/bin/macosx/) and run the `.pkg` installer.


---

###  2. Install RStudio (Posit Desktop)

**Official site:** [Posit RStudio Desktop](https://posit.co/download/rstudio-desktop/)

| OS | Command or Link |
|----|-----------------|
| **Windows** | Download the `.exe` installer and run it. |
| **macOS** | Download the `.dmg` installer and drag RStudio into Applications.

---

###  3. Verify installation
```bash
R --version
Rscript -e 'cat(R.version.string, "\n")'
```
  
</details>
<details>
   <summary><strong>üíª INSTALLATION GUIDE: INDAGO</strong> ‚ñ∏</summary>


## How to install inDAGO from CRAN or GitHub

### Install the Bioconductor dependencies

```r

# Install Bioconductor dependencies if you don't have them yet
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
bioc_pac <- c(
  "XVector",
  "ShortRead",
  "S4Vectors",
  "rtracklayer",
  "Rsubread",
  "Rsamtools",
  "Rfastp",
  "limma",
  "HTSFilter",
  "edgeR",
  "Biostrings",
  "BiocGenerics"
) 
for (pac in bioc_pac) {
  if (!requireNamespace(pac, quietly = TRUE))
    BiocManager::install(pac)
}

```

### Install inDAGO from GitHub

```r
#Install devtools if you don‚Äôt have it yet
if (!requireNamespace("devtools", quietly = TRUE))
  install.packages("devtools")

# Install inDAGO
devtools::install_github("inDAGOverse/inDAGO")

```

### Install inDAGO from CRAN (https://cran.r-project.org/web/packages/inDAGO/index.html)

```r

# Install inDAGO
install.packages("inDAGO")

```
</details>
<details>
   <summary><strong>üöÄ HOW TO LOAD AND LAUNCH THE APP</strong> ‚ñ∏</summary>

```r
# Load and launch the app
library(inDAGO)
inDAGO::inDAGO()

```
</details>
<details>
   <summary><strong>‚öôÔ∏è TIPS FOR A SEAMLESS EXECUTION</strong> ‚ñ∏</summary>
   
To ensure execution during long time-consuming steps such as reference‚Äëbased alignment:

üí§ Disable sleep mode to keep your system active.

üí° Reduce screen brightness to save power.

These simple precautions can help avoid incomplete runs and unnecessary power consumption.

</details>
<details>
   <summary><strong>üë• AUTHORS & ACKNOWLEDGEMENTS</strong> ‚ñ∏</summary>

- **Authors / Creators**  
  - Carmine Fruggiero (c.fruggiero@tigem.it)  
  - Gaetano Aufiero (gaetano.aufiero@unina.it)

- **Designated maintainer for CRAN Repository**  
  - Carmine Fruggiero (c.fruggiero@tigem.it)

- **Project Supervisor**  
  - Nunzio D'Agostino (nunzio.dagostino@unina.it)
</details>
